adapters/beachfront/beachfront.go

```diff
@@ -0,0 +1,486 @@
+		if internalRequest.Imp[i].Video != nil {
+			isVideo = true
+			bidtype = openrtb_ext.BidTypeVideo
+			break
+		}
+	}
+
+	bidResp, err = postprocess(response, externalRequest, internalRequest.ID, isVideo)
+	if err != nil {
+		return nil, []error{fmt.Errorf("Failed to process the beachfront response\n%s", err)}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(BidCapacity)
+
+	var errs []error
+	for _, sb := range bidResp.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+
+			bid := sb.Bid[i]
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: bidtype,
+			})
+		}
+	}
+	return bidResponse, errs
+}
+
+func postprocess(response *adapters.ResponseData, externalRequest *adapters.RequestData, id string, isVideo bool) (openrtb.BidResponse, error) {
+	var beachfrontResp []BeachfrontResponseSlot
+	var openrtbResp openrtb.BidResponse
+	var err error
+
+	if isVideo {
+		// Regular video ad. Beachfront returns video ads in openRTB format (or close enough).
+		if err = json.Unmarshal(response.Body, &openrtbResp); err != nil {
+			return openrtbResp, err
+		}
+		return postprocessVideo(openrtbResp, externalRequest, id)
+	} else {
+		/* Beachfront currently returns banner ads in a sparse format which is just the openRTB seatbid
+		object. It needs to be wrapped up in openrtb format.
+		*/
+		if err = json.Unmarshal(response.Body, &beachfrontResp); err != nil {
+			return openrtbResp, err
+		}
+
+		openrtbResp.ID = id
+		for range beachfrontResp {
+			openrtbResp.SeatBid = append(openrtbResp.SeatBid, openrtb.SeatBid{})
+		}
+
+		return postprocessBanner(openrtbResp, beachfrontResp, id)
+	}
+}
+
+func postprocessBanner(openrtbResp openrtb.BidResponse, beachfrontResp []BeachfrontResponseSlot, id string) (openrtb.BidResponse, error) {
+	r, _ := regexp.Compile("\\\"([0-9]+)")
+	for k, _ := range openrtbResp.SeatBid {
+		openrtbResp.SeatBid[k].Bid = append(openrtbResp.SeatBid[k].Bid, openrtb.Bid{
+			CrID:  fmt.Sprintf("%s", r.FindStringSubmatch(beachfrontResp[k].Adm)[1]),
+			ImpID: beachfrontResp[k].Slot,
+			Price: beachfrontResp[k].Price,
+			ID:    id,
+			AdM:   beachfrontResp[k].Adm,
+			H:     beachfrontResp[k].H,
+			W:     beachfrontResp[k].W,
+		})
+
+		openrtbResp.SeatBid[k].Seat = Seat
+	}
+
+	return openrtbResp, nil
+}
+
+func postprocessVideo(openrtbResp openrtb.BidResponse, externalRequest *adapters.RequestData, id string) (openrtb.BidResponse, error) {
+	var xtrnal BeachfrontVideoRequest
+	var err error
+
+	if err = json.Unmarshal(externalRequest.Body, &xtrnal); err != nil {
+		return openrtbResp, err
+	}
+
+	/* there will only be one seatBid because beachfront only returns a single video ad
+	but if that were to change this should work on all of them:
+	*/
+	for i := range openrtbResp.SeatBid {
+		for j := range openrtbResp.SeatBid[i].Bid {
+			openrtbResp.SeatBid[i].Bid[j].ImpID = xtrnal.Imp[i].ImpId
+			openrtbResp.SeatBid[i].Bid[j].CrID = xtrnal.Imp[i].ImpId
```

Comment:
Whoa, hang on... `CrID` supposed to identify the _creative_ that was returned. `ImpId` is given by pubs to identify the impression.

It's not good for anything if you just overwrite it with the impression ID here.